### ISSUE TITLE
C++: duplicate error with template function redefinition

### ISSUE THREAD

**Issue by pitrou** [sphinx-doc/sphinx#5755]
I was redirected here from the Breathe issue tracker, specifically https://github.com/michaeljones/breathe/issues/406

I have two different implementations of a C++ template method that are selected using the [SFINAE pattern](https://en.cppreference.com/w/cpp/language/sfinae) (implementations omitted below):
```cpp

class BooleanBuilder : public ArrayBuilder {
 public:
  /// \brief Append a sequence of elements in one shot, with a specified nullmap
  /// \param[in] values_begin InputIterator to the beginning of the values
  /// \param[in] values_end InputIterator pointing to the end of the values
  /// \param[in] valid_begin InputIterator with elements indication valid(1)
  ///  or null(0) values
  /// \return Status
  template <typename ValuesIter, typename ValidIter>
  typename std::enable_if<!std::is_pointer<ValidIter>::value, Status>::type AppendValues(
      ValuesIter values_begin, ValuesIter values_end, ValidIter valid_begin) {
     ...
  }

  /// \brief Append a sequence of elements in one shot, with a specified nullmap
  /// \param[in] values_begin InputIterator to the beginning of the values
  /// \param[in] values_end InputIterator pointing to the end of the values
  /// \param[in] valid_begin uint8_t* indication valid(1) or null(0) values.
  ///  nullptr indicates all values are valid.
  /// \return Status
  template <typename ValuesIter, typename ValidIter>
  typename std::enable_if<std::is_pointer<ValidIter>::value, Status>::type AppendValues(
      ValuesIter values_begin, ValuesIter values_end, ValidIter valid_begin) {
     ...
  }
};
```

When trying to integrate Doxygen documentation for that class's members, I get the following error:
```
Warning, treated as error:
/home/antoine/arrow/docs/source/cpp/api/builder.rst:31:Duplicate declaration.
Makefile:71: recipe for target 'html' failed
make: *** [html] Error 2
```

Sphinx version is 1.8.2, Python version is 3.7.

**Comment by jakobandersen** [comment]
Thanks. I wonder how complex a comparison the compiler is supposed to do. As example:
```c++
#include <type_traits>

template<bool B>
typename std::enable_if<B>::type f() {}

template<bool B>
typename std::enable_if<!B>::type f() {}

template<bool B> // not a duplicate until instantiation
typename std::enable_if<!!B>::type f() {}

template<bool A, bool B>
typename std::enable_if<A && B>::type g() {}

template<bool A, bool B> // not a duplicate until instantiation
typename std::enable_if<B && A>::type g() {}

int main() {
	//f<true>(); // ambiguous
	f<false>();
	//g<true, true>(); // ambiguous
}
```
I think the concepts extension comes with new rules for expression comparison, at least for requires clauses.

Fixing this shouldn't be too complicated, though declarations that differ only in the template parameter names will not be found as duplicates any time soon (#2057).

**Comment by jakobandersen** [comment]
Should be fixed in the 1.8 branch now.
